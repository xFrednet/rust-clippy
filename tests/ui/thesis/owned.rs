//@rustc-env: CLIPPY_PETS_PRINT=1

struct Animal {
    legs: u32,
    heads: u32,
}

// Check arguments are correctly detected
#[warn(clippy::borrow_pats)]
fn take_one(_animal: Animal) {}

#[warn(clippy::borrow_pats)]
fn take_two(_animal_1: Animal, _animal_2: Animal) {}

fn take_pair((_animal_1, _animal_2): (Animal, Animal)) {}

#[warn(clippy::borrow_pats)]
fn pat_return_owned_arg(animal: Animal) -> Animal {
    animal
}

#[forbid(clippy::borrow_pats)]
// #[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1(a: String) -> String {
    if !a.is_empty() {
        return a;
    }

    "hey".to_string()
}

// #[forbid(clippy::borrow_pats)]
// #[warn(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_1_test(a: u32) -> u32 {
    if !a.is_power_of_two() {
        return a;
    }

    19
}

// #[forbid(clippy::borrow_pats)]
fn pat_maybe_return_owned_arg_2(a: String) -> String {
    let ret;
    if !a.is_empty() {
        ret = a;
    } else {
        ret = "hey".to_string();
        println!("{a:#?}");
    }
    ret
}

fn pat_copy_to_self(mut animal: Animal) {
    animal.heads = animal.legs;
}

fn main() {}
