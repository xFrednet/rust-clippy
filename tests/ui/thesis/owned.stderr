thread 'rustc' panicked at clippy_lints/src/borrow_pats/owned.rs:191:17:
not yet implemented
stack backtrace:
   0:     0x7dfffab8caf6 - std::backtrace_rs::backtrace::libunwind::trace::hf298c77755f5b12e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7dfffab8caf6 - std::backtrace_rs::backtrace::trace_unsynchronized::h4da6a489ce2db74e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7dfffab8caf6 - std::sys_common::backtrace::_print_fmt::h051beca4a6f5cb98
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7dfffab8caf6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcb86488900881a45
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7dfffabdda7c - core::fmt::rt::Argument::fmt::haf43875fa4a971bb
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/rt.rs:142:9
   5:     0x7dfffabdda7c - core::fmt::write::h44c7f32ae3be5535
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/mod.rs:1120:17
   6:     0x7dfffab814bf - std::io::Write::write_fmt::hc3fec0223f64c010
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/io/mod.rs:1846:15
   7:     0x7dfffab8c8a4 - std::sys_common::backtrace::_print::h7c03376d42fd745a
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7dfffab8c8a4 - std::sys_common::backtrace::print::h9ff10455f68b06c1
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7dfffab8f5eb - std::panicking::default_hook::{{closure}}::he4f9865f1b72b2d5
  10:     0x7dfffab8f339 - std::panicking::default_hook::h3d59f13a05d49fe6
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:292:9
  11:     0x7dfffda731ec - std[19d33715048712a6]::panicking::update_hook::<alloc[6fc58ab69ba746fe]::boxed::Box<rustc_driver_impl[4c01069949ad96dd]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7dfffab8fd50 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf41c577e6ac4b0df
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2030:9
  13:     0x7dfffab8fd50 - std::panicking::rust_panic_with_hook::h70f59ab226ae2c5c
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:786:13
  14:     0x7dfffab8fa59 - std::panicking::begin_panic_handler::{{closure}}::h47fc7126fa960711
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:652:13
  15:     0x7dfffab8cfd6 - std::sys_common::backtrace::__rust_end_short_backtrace::h5b94f863278adeaf
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7dfffab8f7c4 - rust_begin_unwind
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:648:5
  17:     0x7dfffabd9f95 - core::panicking::panic_fmt::h96a33d9be4ce8c0f
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:72:14
  18:     0x7dfffabda053 - core::panicking::panic::hc66a585cf4bd3ee2
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:144:5
  19:     0x64d9bf14b29e - <clippy_lints::borrow_pats::owned::OwnedAnalysis as rustc_middle::mir::visit::Visitor>::visit_assign::hc84e14fb0d3eea88
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/owned.rs:191:17
  20:     0x64d9bf62ba04 - rustc_middle::mir::visit::Visitor::super_statement::h978ac1d7ad2004dc
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:380:25
  21:     0x64d9bf62bd2d - rustc_middle::mir::visit::Visitor::visit_statement::h1bbb6b6e12a17546
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:103:17
  22:     0x64d9bf62cb49 - rustc_middle::mir::visit::Visitor::super_basic_block_data::ha9c57e29847285d7
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/compiler/rustc_middle/src/mir/visit.rs:307:21
  23:     0x64d9bf14af09 - <clippy_lints::borrow_pats::owned::OwnedAnalysis as rustc_middle::mir::visit::Visitor>::visit_basic_block_data::h925a423b9579d672
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/owned.rs:161:9
  24:     0x64d9bf18d993 - clippy_lints::borrow_pats::util::visit_body_in_order::hdd8cd37adf7a4508
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/util.rs:179:9
  25:     0x64d9bf14a9e2 - clippy_lints::borrow_pats::owned::OwnedAnalysis::run::h6a8a187124cd6c1b
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/owned.rs:70:9
  26:     0x64d9bf5bbc2a - <clippy_lints::borrow_pats::BorrowPats as rustc_lint::passes::LateLintPass>::check_body::hf3889422ad231c8d
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats/mod.rs:1046:36
  27:     0x7dfffddfa83b - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_body
  28:     0x7dfffddfb6e4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_fn
  29:     0x7dfffddeaae4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_item
  30:     0x7dffffa2d473 - rustc_lint[a04349e93729da6f]::late::check_crate::{closure#0}
  31:     0x7dffffa2d6d1 - rustc_lint[a04349e93729da6f]::late::check_crate
  32:     0x7dffff8a39c9 - rustc_interface[3868b2f10de3354]::passes::analysis
  33:     0x7dffff8a2669 - rustc_query_impl[80d5b36ddf5440ab]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>
  34:     0x7dffffb33ce5 - rustc_query_system[3c4b6c562f3523cb]::query::plumbing::try_execute_query::<rustc_query_impl[80d5b36ddf5440ab]::DynamicConfig<rustc_query_system[3c4b6c562f3523cb]::query::caches::SingleCache<rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[80d5b36ddf5440ab]::plumbing::QueryCtxt, false>
  35:     0x7dffffb33a49 - rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7dffffb42f59 - rustc_interface[3868b2f10de3354]::interface::run_compiler::<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}
  37:     0x7dffffdbdc45 - std[19d33715048712a6]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>
  38:     0x7dffffdbda72 - <<std[19d33715048712a6]::thread::Builder>::spawn_unchecked_<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#1} as core[a9f526eb09c6e7cd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x7dfffab99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h79a9e4aa91d71737
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  40:     0x7dfffab99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6eabe626cfdc74a5
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  41:     0x7dfffab99725 - std::sys::pal::unix::thread::Thread::new::thread_start::h8428b71c070d7cf2
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys/pal/unix/thread.rs:108:17
  42:     0x7dfff3094ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  43:     0x7dfff3126850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  44:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: please attach the file at `/home/xfrednet/workspace/rust/rust-clippy/rustc-ice-2024-04-15T21_13_54-871759.txt` to your bug report

note: compiler flags: -Z ui-testing -Z deduplicate-diagnostics=no

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.78 (01d878985 2024-04-11)

